lib/Controller/DiscoPower.php

```diff
@@ -0,0 +1,77 @@
+<?php
+
+declare(strict_types=1);
+
+namespace SimpleSAML\Module\discopower\Controller;
+
+use SimpleSAML\Configuration;
+use SimpleSAML\Error;
+use SimpleSAML\Session;
+use SimpleSAML\XHTML\Template;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpFoundation\JsonResponse;
+
+class DiscoPower
+{
+    /**
+     * @param \Symfony\Component\HttpFoundation\Request $request The current request.
+     */
+    public function main(Request $request): void
+    {
+        try {
+            $discoHandler = new \SimpleSAML\Module\discopower\PowerIdPDisco(
+                ['saml20-idp-remote'],
+                'poweridpdisco'
+            );
+        } catch (\Exception $exception) {
+            // An error here should be caused by invalid query parameters
+            throw new \SimpleSAML\Error\Error('DISCOPARAMS', $exception);
+        }
+        
+        try {
+            $discoHandler->handleRequest();
```

Comment:
I think it makes sense to change `handleRequest` to return the template instead of sending it..  The main-controller would then return a Response instead of void..

Also one nitpick;  you've added use-statements but you're still using the full namespaced class names...

Reply:
Because it derives from IdPDisco in simplesamlphp main, this cannot be done until SSP changes that.

Reply:
Ahh, my bad!  We could however wrap this call in a RunnableResponse.